### ISSUE TITLE
Internal error on command:"_BED_MESH_CALIBRATE"

### ISSUE THREAD

**Issue by DerDreschner** [beacon3d/beacon_klipper#18]
I use Klipper with adaptive bed mesh before each print. Right now, the print stopped during calibration with an internal error on command "_BED_MESH_CALIBRATE". The logfile for klippy contains the following error message when this happened:

```
Unhandled exception during run
Traceback (most recent call last):
  File "/home/biqu/klipper/klippy/klippy.py", line 217, in run
    self.reactor.run()
  File "/home/biqu/klipper/klippy/reactor.py", line 292, in run
    g_next.switch()
  File "/home/biqu/klipper/klippy/reactor.py", line 340, in _dispatch_loop
    timeout = self._check_timers(eventtime, busy)
  File "/home/biqu/klipper/klippy/reactor.py", line 158, in _check_timers
    t.waketime = waketime = t.callback(eventtime)
  File "/home/biqu/klipper/klippy/reactor.py", line 48, in invoke
    res = self.callback(eventtime)
  File "/home/biqu/klipper/klippy/extras/beacon.py", line 643, in <lambda>
    self.reactor.register_async_callback(lambda e: self._stream_flush())
  File "/home/biqu/klipper/klippy/extras/beacon.py", line 627, in _stream_flush
    cb(sample)
  File "/home/biqu/klipper/klippy/extras/beacon.py", line 1340, in _handle
    self.cb(sample)
  File "/home/biqu/klipper/klippy/extras/beacon.py", line 1979, in cb
    xi = int(round((x - min_x) / self.step_x))
OverflowError: cannot convert float infinity to integer
Transition to shutdown state: Unhandled exception during run
Dumping gcode input 0 blocks
[...]
b'stepcompress o=8 i=0 c=2 a=0: Invalid sequence'
b'stepcompress o=8 i=0 c=2 a=0: Invalid sequence'
b'stepcompress o=8 i=0 c=2 a=0: Invalid sequence'
b'stepcompress o=8 i=0 c=2 a=0: Invalid sequence'
b'stepcompress o=8 i=0 c=2 a=0: Invalid sequence'
b'stepcompress o=8 i=0 c=2 a=0: Invalid sequence'
Internal error on command:"_BED_MESH_CALIBRATE"
Traceback (most recent call last):
  File "/home/biqu/klipper/klippy/gcode.py", line 211, in _process_commands
    handler(gcmd)
  File "/home/biqu/klipper/klippy/gcode.py", line 137, in <lambda>
    func = lambda params: origfunc(self._get_extended_params(params))
  File "/home/biqu/klipper/klippy/gcode.py", line 137, in <lambda>
    func = lambda params: origfunc(self._get_extended_params(params))
  File "/home/biqu/klipper/klippy/extras/beacon.py", line 1678, in cmd_BED_MESH_CALIBRATE
    self.calibrate(gcmd)
  File "/home/biqu/klipper/klippy/extras/beacon.py", line 1873, in calibrate
    clusters = self._sample_mesh(gcmd, path, speed, runs)
  File "/home/biqu/klipper/klippy/extras/beacon.py", line 2008, in _sample_mesh
    self._fly_path(path, speed, runs)
  File "/home/biqu/klipper/klippy/extras/beacon.py", line 1939, in _fly_path
    self.toolhead.manual_move([x, y, None], speed)
  File "/home/biqu/klipper/klippy/toolhead.py", line 473, in manual_move
    self.move(curpos, speed)
  File "/home/biqu/klipper/klippy/toolhead.py", line 465, in move
    self.lookahead.add_move(move)
  File "/home/biqu/klipper/klippy/toolhead.py", line 187, in add_move
    self.flush(lazy=True)
  File "/home/biqu/klipper/klippy/toolhead.py", line 176, in flush
    self.toolhead._process_moves(queue[:flush_count])
  File "/home/biqu/klipper/klippy/toolhead.py", line 354, in _process_moves
    self._advance_move_time(next_move_time)
  File "/home/biqu/klipper/klippy/toolhead.py", line 312, in _advance_move_time
    self._advance_flush_time(flush_time)
  File "/home/biqu/klipper/klippy/toolhead.py", line 292, in _advance_flush_time
    sg(sg_flush_time)
  File "/home/biqu/klipper/klippy/stepper.py", line 228, in generate_steps
    raise error("Internal error in stepcompress")
stepper.error: Internal error in stepcompress
Internal error on command:"_BED_MESH_CALIBRATE"
Internal error on command:"BED_MESH_CALIBRATE"
Traceback (most recent call last):
  File "/home/biqu/klipper/klippy/gcode.py", line 211, in _process_commands
    handler(gcmd)
  File "/home/biqu/klipper/klippy/gcode.py", line 137, in <lambda>
    func = lambda params: origfunc(self._get_extended_params(params))
  File "/home/biqu/klipper/klippy/extras/gcode_macro.py", line 189, in cmd
    self.template.run_gcode_from_command(kwparams)
  File "/home/biqu/klipper/klippy/extras/gcode_macro.py", line 68, in run_gcode_from_command
    self.gcode.run_script_from_command(self.render(context))
  File "/home/biqu/klipper/klippy/gcode.py", line 226, in run_script_from_command
    self._process_commands(script.split('\n'), need_ack=False)
  File "/home/biqu/klipper/klippy/gcode.py", line 211, in _process_commands
    handler(gcmd)
  File "/home/biqu/klipper/klippy/gcode.py", line 137, in <lambda>
    func = lambda params: origfunc(self._get_extended_params(params))
  File "/home/biqu/klipper/klippy/gcode.py", line 137, in <lambda>
    func = lambda params: origfunc(self._get_extended_params(params))
  File "/home/biqu/klipper/klippy/extras/beacon.py", line 1678, in cmd_BED_MESH_CALIBRATE
    self.calibrate(gcmd)
  File "/home/biqu/klipper/klippy/extras/beacon.py", line 1873, in calibrate
    clusters = self._sample_mesh(gcmd, path, speed, runs)
  File "/home/biqu/klipper/klippy/extras/beacon.py", line 2008, in _sample_mesh
    self._fly_path(path, speed, runs)
  File "/home/biqu/klipper/klippy/extras/beacon.py", line 1939, in _fly_path
    self.toolhead.manual_move([x, y, None], speed)
  File "/home/biqu/klipper/klippy/toolhead.py", line 473, in manual_move
    self.move(curpos, speed)
  File "/home/biqu/klipper/klippy/toolhead.py", line 465, in move
    self.lookahead.add_move(move)
  File "/home/biqu/klipper/klippy/toolhead.py", line 187, in add_move
    self.flush(lazy=True)
  File "/home/biqu/klipper/klippy/toolhead.py", line 176, in flush
    self.toolhead._process_moves(queue[:flush_count])
  File "/home/biqu/klipper/klippy/toolhead.py", line 354, in _process_moves
    self._advance_move_time(next_move_time)
  File "/home/biqu/klipper/klippy/toolhead.py", line 312, in _advance_move_time
    self._advance_flush_time(flush_time)
  File "/home/biqu/klipper/klippy/toolhead.py", line 292, in _advance_flush_time
    sg(sg_flush_time)
  File "/home/biqu/klipper/klippy/stepper.py", line 228, in generate_steps
    raise error("Internal error in stepcompress")
stepper.error: Internal error in stepcompress
Internal error on command:"BED_MESH_CALIBRATE"
Internal error on command:"PRINT_START"
Traceback (most recent call last):
  File "/home/biqu/klipper/klippy/gcode.py", line 211, in _process_commands
    handler(gcmd)
  File "/home/biqu/klipper/klippy/gcode.py", line 137, in <lambda>
    func = lambda params: origfunc(self._get_extended_params(params))
  File "/home/biqu/klipper/klippy/extras/gcode_macro.py", line 189, in cmd
    self.template.run_gcode_from_command(kwparams)
  File "/home/biqu/klipper/klippy/extras/gcode_macro.py", line 68, in run_gcode_from_command
    self.gcode.run_script_from_command(self.render(context))
  File "/home/biqu/klipper/klippy/gcode.py", line 226, in run_script_from_command
    self._process_commands(script.split('\n'), need_ack=False)
  File "/home/biqu/klipper/klippy/gcode.py", line 211, in _process_commands
    handler(gcmd)
  File "/home/biqu/klipper/klippy/gcode.py", line 137, in <lambda>
    func = lambda params: origfunc(self._get_extended_params(params))
  File "/home/biqu/klipper/klippy/extras/gcode_macro.py", line 189, in cmd
    self.template.run_gcode_from_command(kwparams)
  File "/home/biqu/klipper/klippy/extras/gcode_macro.py", line 68, in run_gcode_from_command
    self.gcode.run_script_from_command(self.render(context))
  File "/home/biqu/klipper/klippy/gcode.py", line 226, in run_script_from_command
    self._process_commands(script.split('\n'), need_ack=False)
  File "/home/biqu/klipper/klippy/gcode.py", line 211, in _process_commands
    handler(gcmd)
  File "/home/biqu/klipper/klippy/gcode.py", line 137, in <lambda>
    func = lambda params: origfunc(self._get_extended_params(params))
  File "/home/biqu/klipper/klippy/gcode.py", line 137, in <lambda>
    func = lambda params: origfunc(self._get_extended_params(params))
  File "/home/biqu/klipper/klippy/extras/beacon.py", line 1678, in cmd_BED_MESH_CALIBRATE
    self.calibrate(gcmd)
  File "/home/biqu/klipper/klippy/extras/beacon.py", line 1873, in calibrate
    clusters = self._sample_mesh(gcmd, path, speed, runs)
  File "/home/biqu/klipper/klippy/extras/beacon.py", line 2008, in _sample_mesh
    self._fly_path(path, speed, runs)
  File "/home/biqu/klipper/klippy/extras/beacon.py", line 1939, in _fly_path
    self.toolhead.manual_move([x, y, None], speed)
  File "/home/biqu/klipper/klippy/toolhead.py", line 473, in manual_move
    self.move(curpos, speed)
  File "/home/biqu/klipper/klippy/toolhead.py", line 465, in move
    self.lookahead.add_move(move)
  File "/home/biqu/klipper/klippy/toolhead.py", line 187, in add_move
    self.flush(lazy=True)
  File "/home/biqu/klipper/klippy/toolhead.py", line 176, in flush
    self.toolhead._process_moves(queue[:flush_count])
  File "/home/biqu/klipper/klippy/toolhead.py", line 354, in _process_moves
    self._advance_move_time(next_move_time)
  File "/home/biqu/klipper/klippy/toolhead.py", line 312, in _advance_move_time
    self._advance_flush_time(flush_time)
  File "/home/biqu/klipper/klippy/toolhead.py", line 292, in _advance_flush_time
    sg(sg_flush_time)
  File "/home/biqu/klipper/klippy/stepper.py", line 228, in generate_steps
    raise error("Internal error in stepcompress")
stepper.error: Internal error in stepcompress
Internal error on command:"PRINT_START"
virtual_sdcard dispatch
Traceback (most recent call last):
  File "/home/biqu/klipper/klippy/extras/virtual_sdcard.py", line 264, in work_handler
    self.gcode.run_script(line)
  File "/home/biqu/klipper/klippy/gcode.py", line 229, in run_script
    self._process_commands(script.split('\n'), need_ack=False)
  File "/home/biqu/klipper/klippy/gcode.py", line 211, in _process_commands
    handler(gcmd)
  File "/home/biqu/klipper/klippy/gcode.py", line 137, in <lambda>
    func = lambda params: origfunc(self._get_extended_params(params))
  File "/home/biqu/klipper/klippy/extras/gcode_macro.py", line 189, in cmd
    self.template.run_gcode_from_command(kwparams)
  File "/home/biqu/klipper/klippy/extras/gcode_macro.py", line 68, in run_gcode_from_command
    self.gcode.run_script_from_command(self.render(context))
  File "/home/biqu/klipper/klippy/gcode.py", line 226, in run_script_from_command
    self._process_commands(script.split('\n'), need_ack=False)
  File "/home/biqu/klipper/klippy/gcode.py", line 211, in _process_commands
    handler(gcmd)
  File "/home/biqu/klipper/klippy/gcode.py", line 137, in <lambda>
    func = lambda params: origfunc(self._get_extended_params(params))
  File "/home/biqu/klipper/klippy/extras/gcode_macro.py", line 189, in cmd
    self.template.run_gcode_from_command(kwparams)
  File "/home/biqu/klipper/klippy/extras/gcode_macro.py", line 68, in run_gcode_from_command
    self.gcode.run_script_from_command(self.render(context))
  File "/home/biqu/klipper/klippy/gcode.py", line 226, in run_script_from_command
    self._process_commands(script.split('\n'), need_ack=False)
  File "/home/biqu/klipper/klippy/gcode.py", line 211, in _process_commands
    handler(gcmd)
  File "/home/biqu/klipper/klippy/gcode.py", line 137, in <lambda>
    func = lambda params: origfunc(self._get_extended_params(params))
  File "/home/biqu/klipper/klippy/gcode.py", line 137, in <lambda>
    func = lambda params: origfunc(self._get_extended_params(params))
  File "/home/biqu/klipper/klippy/extras/beacon.py", line 1678, in cmd_BED_MESH_CALIBRATE
    self.calibrate(gcmd)
  File "/home/biqu/klipper/klippy/extras/beacon.py", line 1873, in calibrate
    clusters = self._sample_mesh(gcmd, path, speed, runs)
  File "/home/biqu/klipper/klippy/extras/beacon.py", line 2008, in _sample_mesh
    self._fly_path(path, speed, runs)
  File "/home/biqu/klipper/klippy/extras/beacon.py", line 1939, in _fly_path
    self.toolhead.manual_move([x, y, None], speed)
  File "/home/biqu/klipper/klippy/toolhead.py", line 473, in manual_move
    self.move(curpos, speed)
  File "/home/biqu/klipper/klippy/toolhead.py", line 465, in move
    self.lookahead.add_move(move)
  File "/home/biqu/klipper/klippy/toolhead.py", line 187, in add_move
    self.flush(lazy=True)
  File "/home/biqu/klipper/klippy/toolhead.py", line 176, in flush
    self.toolhead._process_moves(queue[:flush_count])
  File "/home/biqu/klipper/klippy/toolhead.py", line 354, in _process_moves
    self._advance_move_time(next_move_time)
  File "/home/biqu/klipper/klippy/toolhead.py", line 312, in _advance_move_time
    self._advance_flush_time(flush_time)
  File "/home/biqu/klipper/klippy/toolhead.py", line 292, in _advance_flush_time
    sg(sg_flush_time)
  File "/home/biqu/klipper/klippy/stepper.py", line 228, in generate_steps
    raise error("Internal error in stepcompress")
stepper.error: Internal error in stepcompress
Exiting SD card print (position 34980)
[...]
b'stepcompress o=5 i=0 c=4 a=0: Invalid sequence'
b'stepcompress o=5 i=0 c=4 a=0: Invalid sequence'
Exception in flush_handler
Traceback (most recent call last):
  File "/home/biqu/klipper/klippy/toolhead.py", line 427, in _flush_handler
    self._flush_lookahead()
  File "/home/biqu/klipper/klippy/toolhead.py", line 357, in _flush_lookahead
    self.lookahead.flush()
  File "/home/biqu/klipper/klippy/toolhead.py", line 176, in flush
    self.toolhead._process_moves(queue[:flush_count])
  File "/home/biqu/klipper/klippy/toolhead.py", line 354, in _process_moves
    self._advance_move_time(next_move_time)
  File "/home/biqu/klipper/klippy/toolhead.py", line 312, in _advance_move_time
    self._advance_flush_time(flush_time)
  File "/home/biqu/klipper/klippy/toolhead.py", line 292, in _advance_flush_time
    sg(sg_flush_time)
  File "/home/biqu/klipper/klippy/stepper.py", line 228, in generate_steps
    raise error("Internal error in stepcompress")
stepper.error: Internal error in stepcompress
```

It's the first time something like this happens. My printer.cfg contains the following settings regarding the beacon and bed mesh calibration:

```
[beacon]
serial: /dev/serial/by-id/usb-Beacon_Beacon_RevH_26CEFCBB5154364134202020FF141809-if00
x_offset: -5.22 # update with offset from nozzle on your machine
y_offset: -35.08 # update with offset from nozzle on your machine
mesh_main_direction: y
mesh_runs: 2

[resonance_tester]
accel_chip: beacon
probe_points: 90, 90, 20

...

[bed_mesh]
speed: 175
mesh_min: 15, 45
mesh_max: 285, 255
probe_count: 100, 100
algorithm: bicubic
```

I use a Beacon Rev. H (git version v1.0.0-23-g9d062568) together with a Manta M8P V2.0 board.

**Comment by cinusik** [comment]
Same here, going back two version fixed problem. Pleas fix this :)

**Comment by dalegaard** [comment]
Hi guys,

Thanks for reporting this. A subtle memory management issue snuck in with a performance optimization in 0a12243. It's been fixed in cb35f9c. Sorry for the inconvenience!

Best regards,
Lasse
